core/src/cache.rs

```diff
@@ -688,69 +688,69 @@ impl Cache {
-                    // won't change the state to `EntryState::ImportsResolved` (and for a good
-                    // reason: we wouldn't even know what are the pending imports to resolve). The
-                    // Nickel pipeline should however fail if `resolve_imports` failed at some
-                    // point, anyway.
-                    (Vec::new(), Vec::new())
                 };
 
-                Ok(CacheOp::Done((pending, errors)))
+                // unwrap!(): we called `unwrap()` at the beggining of the enclosing if branch
+                // on the result of `self.terms.get(&file_id)`. We only made recursive calls to
+                // `resolve_imports` in between, which don't remove anything from `self.terms`.
+                let cached_term = self.terms.get_mut(&file_id).unwrap();
+                cached_term.term = transformed_term;
+                cached_term.state = EntryState::ImportsResolving;
+
+                let mut done = Vec::new();
+
+                // Transitively resolve the imports, and accumulate the ids of the resolved
+                // files along the way.
+                for id in pending {
+                    if let CacheOp::Done((mut done_local, _)) = self.resolve_imports(id)? {
+                        done.push(id);
+                        done.append(&mut done_local)
+                    }
+                }
+
+                self.update_state(file_id, EntryState::ImportsResolved);
+
+                Ok(CacheOp::Done((done, import_errors)))
             }
-            _ => Err(CacheError::NotParsed),
+            // [transitory_entry_state]:
+            //
+            // This case is triggered by a cyclic import. The entry is already
+            // being treated by an ongoing call to `resolve_import` higher up in
+            // the call chain, so we don't do anything here.
+            //
+            // Note that in some cases, this intermediate state can be observed by an
+            // external caller: if a first call to `resolve_imports` fails in the middle of
+            // resolving the transitive imports, the end state of the entry is
+            // `ImportsResolving`. Subsequent calls to `resolve_imports` will succeed, but
+            // won't change the state to `EntryState::ImportsResolved` (and for a good
+            // reason: we wouldn't even know what are the pending imports to resolve). The
+            // Nickel pipeline should however fail if `resolve_imports` failed at some
+            // point, anyway.
+            Some(EntryState::ImportsResolving) => Ok(CacheOp::Done((Vec::new(), Vec::new()))),
+            // >= EntryState::ImportsResolved
+            Some(
+                EntryState::ImportsResolved
+                | EntryState::Typechecking
+                | EntryState::Typechecked
+                | EntryState::Transforming
+                | EntryState::Transformed,
```

Comment:
I believe one case such as `Some(state) if state >= EntryState::ImportResolved` is equally clear and a bit less verbose. Enumerating the cases has the advantages that if we ever add a phase, this forces us to update this part of this code (and thus to think about what we should do), but I fee like the ordering is enough here: whatever are the next phases, if we're past `ImportResovled`, there's nothing to do.

I really don't feel strongly about it though, feel free to ignore

Reply:
I agree that would be better. The disadvantage is that then rust can't tell the pattern is exhaustive, so we have to either add a `_ => unreachable!()` or change `None` to a wildcard match. Before, `None` was a wildcard match, and that was one of the things that confused me. ("what can this be besides None?? Oh, it can't.")

Reply:
Oh, I see. Haven't thought about that. If it confused you, it's a sign that it might be better indeed (given than this version is more verbose but not unclear).